compiler/dyno/include/chpl/framework/ErrorWriter.h

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright 2021-2022 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef CHPL_FRAMEWORK_ERRORWRITER_H
+#define CHPL_FRAMEWORK_ERRORWRITER_H
+
+#include <ostream>
+
+#include "chpl/framework/ErrorBase.h"
+#include "chpl/uast/all-uast.h"
+#include "chpl/parsing/parsing-queries.h"
+#include "chpl/util/terminal.h"
+
+namespace chpl {
+
+class Context;
+
+namespace errordetail {
+
+inline Location locate(Context* context, const ID& id) {
+  if (!context) return Location();
+  return parsing::locateId(context, id);
+}
+inline Location locate(Context* context, const uast::AstNode* node) {
+  return locate(context, node->id());
+}
+
+/// \cond DO_NOT_DOCUMENT
+template <typename T>
+struct AsFileName {
+  T t; /* The thing to locate */
+
+  Location location(Context* context) const {
+    return locate(context, t);
+  }
+};
+/// \endcond
+
+} // end namespace errordetail
+
+template <typename T>
+errordetail::AsFileName<T> fileNameOf(T t) {
+  return errordetail::AsFileName<T> { std::move(t) };
+}
+
+class ErrorWriter {
```

Comment:
This class needs a docs comment to describe its purpose & we also need docs comments for the methods. We need to be able to find docs for it when implementing new errors.